File: archive/tools-teacher/app/src/main/java/com/instructure/teacheraid/util/ViewUtils.java
/*
 * Copyright (C) 2016 - present  Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.instructure.teacheraid.util;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.res.Resources;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.Display;

import java.lang.reflect.Method;

public class ViewUtils {

    public static float convertDipsToPixels(float dp, Context context){
        Resources resources = context.getResources();
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, resources.getDisplayMetrics());
    }


    @SuppressLint("NewApi")
    public static int getWindowHeight(Activity activity) {
        DisplayMetrics metrics = new DisplayMetrics();
        Display display = activity.getWindowManager().getDefaultDisplay();

        try {
            // For JellyBean 4.2 (API 17) and onward
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN_MR1) {
                display.getRealMetrics(metrics);
                return metrics.heightPixels;
            } else {
                Method mGetRawHeight = Display.class.getMethod("getRawHeight");
                return (Integer) mGetRawHeight.invoke(display);
            }
        } catch (Exception e) {
            // if neither of the above work, we should at least return something
            // that's fairly close, even if it doesn't account for the navigation bar
            // see: http://stackoverflow.com/questions/10991194/android-displaymetrics-returns-incorrect-screen-size-in-pixels-on-ics
            display.getMetrics(metrics);
            return metrics.heightPixels;
        }
    }

}

File: automation/espresso/src/main/kotlin/com/instructure/espresso/ViewUtils.kt
//
// Copyright (C) 2018-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package com.instructure.espresso

import androidx.test.espresso.util.HumanReadables
import android.view.View

object ViewUtils {

    fun toString(view: View): String {
        return HumanReadables.getViewHierarchyErrorMessage(view, null, "", null)
    }
}
